### PR TITLE
Sharding: async migration to move to sharded schema

### PR DESCRIPTION
**Pull request by macobo**
## Changes

Part of https://github.com/PostHog/posthog/issues/8652

This migration migrates all tables to their replicated counterparts. See the code itself for a longer description as to what's going on.

@yakkomajuri I've tagged you as a reviewer due to this being an async migration. Let me know if you don't have the capacity to review this though.

## How did you test this code?

Unit tests + running the migration many times on fresh setups.

### CODE REVIEW

File: posthog/async_migrations/test/test_0004_replicated_schema.py
Comment (hazzadous) [inline posthog/async_migrations/test/test_0004_replicated_schema.py:92]: Can we be explicit about the table engine checking here? e.g. 

```
assert dict(table_engines) == {
    "sharded_events": "Replicated...",
    ...
}
```

Comment (macobo) [reply posthog/async_migrations/test/test_0004_replicated_schema.py:92]: The snapshot test below achieves this

File: ee/clickhouse/sql/cohort.py
Comment (macobo) [reply ee/clickhouse/sql/cohort.py:9]: No. All these schema statements need to be lazy to account for CLICKHOUSE_REPLICATION. In this case CLICKHOUSE_REPLICATION influences the table engine (Replicated vs non-replicated).

See https://github.com/PostHog/posthog/issues/8652 for the full plan of the sharded migration - we'll remove CLICKHOUSE_REPLICATION at a later step.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (hazzadous) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:46]: Can we test for this?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (hazzadous) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:47]: It looks like this isn't something I, as someone running the migration has to worry about? The migration will try to stop any merge? Does the stop operation block until any existing merges have completed?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (hazzadous) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:45]: Is this enforced in anyway? What happens if we run concurrently?

Comment (hazzadous) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:45]: Oh, wait maybe I misunderstood, this is one clickhouse node?

Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:45]: It is in #validate_number_of_nodes_in_cluster

Note this is about clickhouse cluster size not parallel celery processes

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:47]: The operator should not need to know this detail and we:

1. Stop merges in #operations
2. Verify if there are no ongoing merges for each table before moving data.

File: ee/clickhouse/sql/table_engines.py
Comment (hazzadous) [inline ee/clickhouse/sql/table_engines.py:41]: What's this for? Can we add a comment?

File: posthog/async_migrations/test/test_0004_replicated_schema.py
Comment (hazzadous) [inline posthog/async_migrations/test/test_0004_replicated_schema.py:73]: Is it worth validating that the other operations work as expected, e.g. writing to `writable_events` etc? I'm not sure how much of that would already be covered by the rest of the test suite having run.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (hazzadous) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:134]: If we are backing up, should we also verify that the backups are valid?

Comment (hazzadous) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:134]: (I have no idea if that is a valid suggestion tbh but thought worth asking, non-blocking)

Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:134]: Not sure what you mean by that. I'm just leaving the old table (with the previous schema) around with `backup_table_name`. Do you have suggestions around the naming?

File: posthog/async_migrations/test/test_0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/test/test_0004_replicated_schema.py:73]: I don't think there's value on this.

On the pragmatical side, we won't be able to cover EVERYTHING_POSTHOG_IS_ABLE_TO_DO in a single test suite.

On the theoretical side, further down the line of this project we'll get rid of CLICKHOUSE_REPLICATED and the whole test suite will be testing those write paths.

Comment (hazzadous) [inline posthog/async_migrations/test/test_0004_replicated_schema.py:73]: Also, may be worth checking e.g. that merges have been started.

**Review by hazzadous — APPROVED** [review]
I _think_ it looks ok, but another pair of eyes wouldn't go amiss.

I've left some comments but I think non of them are issues that should stop moving forward.

File: posthog/async_migrations/test/test_0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/test/test_0004_replicated_schema.py:73]: Not sure how - didn't find anything in [ClickHouse docs](https://clickhouse.com/docs/en/sql-reference/statements/system/)

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:46]: Yep! Will add this.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (hazzadous) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:134]: I'm asking here if we could add tests to verify backup tables are present

Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:134]: Ah I see! Done!

File: ee/clickhouse/sql/table_engines.py
Comment (macobo) [reply ee/clickhouse/sql/table_engines.py:42]: Done!

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py]: Improved the comment

**Review by guidoiaquinti — APPROVED** [review]

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:95]: also precheck?

**Review by yakkomajuri — COMMENTED** [review]
quick comments from me - mostly I'd love to see a precheck and a healthcheck.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:94]: Tell me, what is the precheck and what belongs there?

**Comment by macobo** [comment]
@yakkomajuri Re `healthcheck` - why? What would be the proper healthcheck for this migration?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:118]: All operations need to be resumable, so you need a if not exists here.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:135]: Both of these need to be resumable as you pointed out yourself in slack. Specifically for the case, when someone restarts the migration from scratch. 
Not sure what the best way is exactly, but e.g. you can add a verification to pre-check that the backup table names aren't used.
For rollback: it shouldn't fail if the op wasn't ran & it should be resumable, so if the backup table exists rename, otherwise no-op?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:126]: Are these resumable?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:140]: Is `create_materialized_view` resumable?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:94]: precheck = verifications we ran before starting the migration (& when someone restarts after having had a rollback)

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:118]: Define resumability?

If you mean that it should be _idempotent_ - nope, it should not be. We need to ensure that the schemas of the two tables are _exactly_ identical, which the `AS` part of the query enforces. If the table exists but with a different schema, we have bigger problems.

Note that we drop the table on the rollback side of things. Though the issue around dropping tables/schema alterations still exists - I think we should roll back starting operation-1 not operation.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:135]: Again - define resumability?

Both of these are atomic operations, just updating metadata. The list of things that can go wrong is hence limited - e.g. clickhouse being down or naming conflicts. First we can't avoid and rollbacks handled if we started from op-1, the latter is being tested for.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:205]: plugin server does not communicate with clickhouse at all.

These operations are atomic and don't actually copy any data - see the long migration background at the top. Instead it creates a hard link and then `rm -rf`s the right directory. Neither should be a long-running operation.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:140]: Again, define resumability.

We don't need to handle the case where the table exists because we explicitly drop it above. If we have concurrent schema modifications going on, we have a lot of other worries as well.

**Comment by macobo** [comment]
Summary on the discussion: We should merge the PR to reduce WIP on my end. @tiina303 or @yakkomajuri will take a last look at the PR to tag everything.

To follow up on separately:
- zookeeper paths are an issue. We need to handle that.
- Rename table is not rollbackable if it fails but not really. Re-implement this in a way that makes it so.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:100]: What happens if they have multiple nodes? What should they do?

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:136]: discussed in sync will be follow-up

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:145]: discussed in sync will be follow-up

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:150]: discussed in sync will be follow-up

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:185]: sth to consider: if this fails we won't get a very nice error for the user probably, which is why I added the try catch and log sql too into execute_clickhouse_op etc.

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:212]: Do we see info level logs by default?

**Review by tiina303 — APPROVED** [review]
Only looked at the migration

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:94]: Not part of this step. Context in https://github.com/PostHog/posthog/issues/8652

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:212]: yes

File: posthog/async_migrations/migrations/0004_replicated_schema.py
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:185]: I highlighted this in my feedback - execute_op_* is not a good API and wouldn't work here without significant refactoring and also getting rid of it requires also changing other migrations.

Comment (tiina303) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:185]: you don't have to use `execute_op` but it could be nice to check that we give users/us enough info to debug why this step failed

**Pull request merged**
